File: lib/fulink/superAgiV7.ts
/**
 * SuperAgi v7 포팅 (assets/lib core_ai/super_agi_v7.dart)
 * 반응구간 + 스탑헌팅 밴드 + EV + 동적 레버/수량/TP·SL
 * - FuState 대신 최소 입력(캔들, 방향, 진입가, 목표가, 지지/저항) 사용
 */

import type { Candle } from '@/types';

export type SuperAgiV7Input = {
  candles: Candle[];
  finalDir: 'LONG' | 'SHORT' | 'WATCH';
  entry: number;
  target: number;
  reactLow?: number;
  reactHigh?: number;
  s1?: number;  // support
  r1?: number;  // resistance
  signalProb?: number;  // 0..100
  confidence?: number;  // 0..100
  livePrice: number;
  accountUsdt?: number;
  riskPct?: number;
};

export type SuperAgiV7Out = {
  state: string;
  evR: number;
  stopHuntRisk: number;
  huntBandLow: number;
  huntBandHigh: number;
  slRecommended: number;
  qty: number;
  leverage: number;
  tp1: number;
  tp2: number;
  tp3: number;
  managerLine1: string;
  managerLine2: string;
};

function atr14(candles: Candle[]): number {
  if (candles.length < 2) return 0;
  const n = Math.min(14, candles.length - 1);
  let sum = 0;
  for (let i = candles.length - n; i < candles.length; i++) {
    if (i <= 0) continue;
    const hi = candles[i]!.high;
    const lo = candles[i]!.low;
    const prevClose = candles[i - 1]!.close;
    const tr = Math.max(hi - lo, Math.abs(hi - prevClose), Math.abs(lo - prevClose));
    sum += tr;
  }
  return sum / n;
}

function minLow(candles: Candle[]): number {
  if (!candles.length) return 0;
  let m = candles[0]!.low;
  for (const c of candles) {
    if (c.low < m) m = c.low;
  }
  return m;
}

function maxHigh(candles: Candle[]): number {
  if (!candles.length) return 0;
  let m = candles[0]!.high;
  for (const c of candles) {
    if (c.high > m) m = c.high;
  }
  return m;
}

function swingRange(candles: Candle[]): [number, number] {
  if (candles.length < 5) return [minLow(candles), maxHigh(candles)];
  const start = Math.max(2, candles.length - 50);
  let swingLow = Infinity;
  let swingHigh = -Infinity;
  for (let i = start; i < candles.length - 2; i++) {
    const l = candles[i]!.low;
    const h = candles[i]!.high;
    if (l < candles[i - 1]!.low && l < candles[i - 2]!.low && l < candles[i + 1]!.low && l < candles[i + 2]!.low) {
      swingLow = Math.min(swingLow, l);
    }
    if (h > candles[i - 1]!.high && h > candles[i - 2]!.high && h > candles[i + 1]!.high && h > candles[i + 2]!.high) {
      swingHigh = Math.max(swingHigh, h);
    }
  }
  if (swingLow === Infinity) swingLow = minLow(candles);
  if (swingHigh === -Infinity) swingHigh = maxHigh(candles);
  return [swingLow, swingHigh];
}

function wickCluster(candles: Candle[]): [number, number, number] {
  if (!candles.length) return [0, 0, 0];
  const start = Math.max(0, candles.length - 60);
  const slice = candles.slice(start);
  let wickiness = 0;
  const lows: number[] = [];
  const highs: number[] = [];
  for (const k of slice) {
    lows.push(k.low);
    highs.push(k.high);
    const range = Math.max(1e-9, Math.abs(k.high - k.low));
    const lowerWick = Math.max(0, Math.min(k.open, k.close) - k.low);
    const upperWick = Math.max(0, k.high - Math.max(k.open, k.close));
    wickiness += (lowerWick + upperWick) / range;
  }
  wickiness = Math.max(0, Math.min(2, wickiness / slice.length));
  lows.sort((a, b) => a - b);
  highs.sort((a, b) => a - b);
  const wickLow = lows[Math.floor(lows.length * 0.1)] ?? lows[0] ?? 0;
  const wickHigh = highs[Math.floor(highs.length * 0.9)] ?? highs[highs.length - 1] ?? 0;
  return [wickLow, wickHigh, wickiness / 2];
}

function zoneTouchScore(candles: Candle[], zLow: number, zHigh: number): number {
  if (!candles.length) return 0;
  const start = Math.max(0, candles.length - 30);
  const slice = candles.slice(start);
  let touches = 0;
  for (const k of slice) {
    if (k.low <= zHigh && k.high >= zLow) touches++;
  }
  return Math.max(0, Math.min(1, touches / slice.length));
}

const kAtr = 1.0;
const kZone = 0.2;

export function computeSuperAgiV7(input: SuperAgiV7Input): SuperAgiV7Out {
  const {
    candles,
    finalDir,
    entry: entryInput,
    target: targetInput,
    reactLow: rLow,
    reactHigh: rHigh,
    s1,
    r1,
    signalProb = 50,
    confidence = 50,
    livePrice,
    accountUsdt = 10000,
    riskPct = 1,
  } = input;

  const seed = Math.max(1, Math.min(1e12, accountUsdt));
  const risk = Math.max(0.001, Math.min(0.5, riskPct / 100));
  const riskMoney = seed * risk;

  const price = livePrice;
  const zLow = (rLow != null && rLow > 0) ? rLow : Math.min(s1 ?? price, price);
  const zHigh = (rHigh != null && rHigh > 0) ? rHigh : Math.max(r1 ?? price, price);
  const zWidth = Math.max(1e-9, Math.abs(zHigh - zLow));

  const atr = atr14(candles);
  const buffer = Math.max(atr * kAtr, zWidth * kZone);

  const [swingLow, swingHigh] = swingRange(candles);
  const [wickLow, wickHigh, wickiness] = wickCluster(candles);
  const extremeLow = minLow(candles);
  const extremeHigh = maxHigh(candles);

  const huntLow = Math.min(swingLow, wickLow, extremeLow) - buffer;
  const huntHigh = Math.max(swingHigh, wickHigh, extremeHigh) + buffer;

  const dir = finalDir === 'WATCH' ? 'WATCH' : finalDir;
  const slRec = dir === 'SHORT' ? huntHigh : huntLow;

  const entry = entryInput > 0 ? entryInput : livePrice;
  const stopDist = Math.max(1e-9, Math.abs(entry - slRec));
  const qty = Math.max(0, Math.min(1e12, riskMoney / stopDist));
  const notional = qty * entry;
  const lev = seed > 0 ? Math.min(999, notional / seed) : 0;

  const target = targetInput > 0 ? targetInput : (dir === 'SHORT' ? entry - 3 * stopDist : entry + 3 * stopDist);
  const oneR = stopDist;
  const tp1 = entry + (target - entry) * 0.4;
  const tp2 = entry + (target - entry) * 0.75;
  const tp3 = target;

  const p = Math.max(0.05, Math.min(0.95, (signalProb / 100) || (confidence / 100)));
  const winR = Math.min(50, Math.abs(tp3 - entry) / stopDist);
  const evR = p * winR - (1 - p) * 1;

  const zoneTouches = zoneTouchScore(candles, zLow, zHigh);
  const atrVsWidth = Math.min(5, atr / (zWidth + 1e-9));
  const stopHuntRisk = Math.max(0, Math.min(100, Math.round(wickiness * 55 + zoneTouches * 35 + atrVsWidth * 10)));

  const inside = price >= zLow && price <= zHigh;
  let state = 'WAIT';
  if (stopHuntRisk >= 70) state = 'LOCK';
  else if (inside) {
    const vNow = candles.length ? candles[candles.length - 1]!.volume : 0;
    const volSlice = candles.slice(-20);
    const vAvg = volSlice.length ? volSlice.reduce((a, c) => a + c.volume, 0) / volSlice.length : 0;
    const volOk = vAvg > 0 ? vNow / vAvg >= 1.25 : false;
    if (volOk && p >= 0.55) state = 'CONFIRM';
    else state = 'TEST';
  }

  const stateLabel = (s: string) => {
    switch (s) {
      case 'LOCK': return 'LOCK';
      case 'CONFIRM': return '확정';
      case 'TEST': return '테스트';
      case 'FAIL': return '실패';
      default: return '대기';
    }
  };
  const dirLabel = (d: string) => d === 'LONG' ? '롱' : d === 'SHORT' ? '숏' : '관망';

  const evTxt = `${evR >= 0 ? '+' : ''}${evR.toFixed(2)}R`;
  const line1 = `${stateLabel(state)} · ${dirLabel(dir)} · EV ${evTxt}`;
  const line2 = `헌팅위험 ${stopHuntRisk} · SL ${slRec.toFixed(0)} · 레버 ${lev.toFixed(1)}x`;

  return {
    state,
    evR,
    stopHuntRisk,
    huntBandLow: huntLow,
    huntBandHigh: huntHigh,
    slRecommended: slRec,
    qty,
    leverage: lev,
    tp1,
    tp2,
    tp3,
    managerLine1: line1,
    managerLine2: line2,
  };
}

File: assets/lib/core_ai/super_agi_v7.dart

import 'dart:math' as math;

import '../models/fu_state.dart';
import '../app_settings.dart';

/// SUPER AGI v7 (핵심: 반응구간 + 스탑헌팅 밴드 + EV + 동적 레버)
/// - 기존 엔진/기능 삭제 없이 "추가"로 계산해 UI/브리핑에 사용
class SuperAgiV7Out {
  final String state; // WAIT/TEST/CONFIRM/FAIL/LOCK
  final double evR; // 기대값 (R 단위)
  final int stopHuntRisk; // 0~100
  final double huntBandLow;
  final double huntBandHigh;
  final double slRecommended;
  final double qty;
  final double leverage;
  final double tp1;
  final double tp2;
  final double tp3;
  final String managerLine1;
  final String managerLine2;

  const SuperAgiV7Out({
    required this.state,
    required this.evR,
    required this.stopHuntRisk,
    required this.huntBandLow,
    required this.huntBandHigh,
    required this.slRecommended,
    required this.qty,
    required this.leverage,
    required this.tp1,
    required this.tp2,
    required this.tp3,
    required this.managerLine1,
    required this.managerLine2,
  });
}

class SuperAgiV7 {
  /// k1: ATR 계수, k2: zoneWidth 계수
  final double kAtr;
  final double kZone;

  const SuperAgiV7({this.kAtr = 1.0, this.kZone = 0.20});

  SuperAgiV7Out compute({
    required FuState s,
    required double livePrice,
    double? seedUsdt,
  }) {
    final seed = (seedUsdt ?? AppSettings.accountUsdt).clamp(1.0, 1e12);
    final riskPct = AppSettings.riskPct.clamp(0.1, 50.0) / 100.0;
    final riskMoney = seed * riskPct;

    // 1) 반응구간(기존 엔진 산출)
    final zLow = (s.reactLow > 0 ? s.reactLow : math.min(s.s1, s.price)).toDouble();
    final zHigh = (s.reactHigh > 0 ? s.reactHigh : math.max(s.r1, s.price)).toDouble();
    final zWidth = (zHigh - zLow).abs().clamp(1e-9, 1e18);

    // 2) ATR(간단 14)
    final atr = _atr14(s.candles);

    // 3) 스탑헌팅 밴드 (구조 + 꼬리 + 안전버퍼)
    final buffer = math.max(atr * kAtr, zWidth * kZone);

    final (swingLow, swingHigh) = _swingRange(s.candles);
    final (wickLow, wickHigh, wickiness) = _wickCluster(s.candles);

    // 유동성 스윕 후보(간단: 최근 저/고 extremes)
    final extremeLow = _minLow(s.candles);
    final extremeHigh = _maxHigh(s.candles);

    // 헌팅 밴드
    final huntLow = math.min(math.min(swingLow, wickLow), extremeLow) - buffer;
    final huntHigh = math.max(math.max(swingHigh, wickHigh), extremeHigh) + buffer;

    // 4) 추천 SL: 방향에 따라 바깥
    final dir = s.finalDir.toUpperCase(); // LONG/SHORT/WATCH
    final slRec = dir == 'SHORT' ? huntHigh : huntLow;

    // 5) 동적 수량/레버 계산 (5% 리스크 고정)
    final entry = (s.entry > 0 ? s.entry : livePrice);
    final stopDist = (entry - slRec).abs().clamp(1e-9, 1e18);
    final qty = (riskMoney / stopDist).clamp(0.0, 1e12);
    final notional = qty * entry;
    final lev = (seed > 0 ? notional / seed : 0.0).clamp(0.0, 999.0);

    // 6) 목표(TP1/2/3): 기존 target이 있으면 그 구간 내 분할, 없으면 1R/2R/3R 구조
    final oneR = stopDist; // 가격 기준 1R 거리
    final target = s.target > 0 ? s.target : (dir == 'SHORT' ? entry - 3.0 * oneR : entry + 3.0 * oneR);
    final tp1 = entry + (target - entry) * 0.40;
    final tp2 = entry + (target - entry) * 0.75;
    final tp3 = target;

    // 7) P(win) 근사 + EV(R)
    final p = _pWin(s);
    final winR = ((tp3 - entry).abs() / stopDist).clamp(0.0, 50.0);
    final evR = (p * winR) - ((1.0 - p) * 1.0);

    // 8) StopHuntRisk(0~100): 꼬리/휘둘림 + zone 근접 횟수 + 구조/변동성
    final zoneTouches = _zoneTouchScore(s.candles, zLow, zHigh);
    final atrVsWidth = (atr / (zWidth + 1e-9)).clamp(0.0, 5.0);
    int risk = (wickiness * 55.0 + zoneTouches * 35.0 + atrVsWidth * 10.0).round().clamp(0, 100);

    // 9) 상태 머신
    final state = _stateMachine(s, livePrice, zLow, zHigh, risk);

    // 10) 매니저 2줄 브리핑
    final evTxt = '${evR >= 0 ? '+' : ''}${evR.toStringAsFixed(2)}R';
    final riskTxt = '헌팅위험 ${risk}';
    final levTxt = '레버 ${lev.toStringAsFixed(1)}x';
    final slTxt = 'SL ${slRec.toStringAsFixed(0)}';
    final line1 = '${stateLabel(state)} · ${dirLabel(dir)} · EV $evTxt';
    final line2 = '${riskTxt} · ${slTxt} · ${levTxt}';

    return SuperAgiV7Out(
      state: state,
      evR: evR,
      stopHuntRisk: risk,
      huntBandLow: huntLow,
      huntBandHigh: huntHigh,
      slRecommended: slRec,
      qty: qty,
      leverage: lev,
      tp1: tp1,
      tp2: tp2,
      tp3: tp3,
      managerLine1: line1,
      managerLine2: line2,
    );
  }

  static String stateLabel(String s) {
    switch (s) {
      case 'LOCK':
        return 'LOCK';
      case 'CONFIRM':
        return '확정';
      case 'TEST':
        return '테스트';
      case 'FAIL':
        return '실패';
      default:
        return '대기';
    }
  }

  static String dirLabel(String dir) {
    switch (dir) {
      case 'LONG':
        return '롱';
      case 'SHORT':
        return '숏';
      default:
        return '관망';
    }
  }

  static String _stateMachine(FuState s, double price, double zLow, double zHigh, int stopHuntRisk) {
    if (s.locked) return 'LOCK';
    if (stopHuntRisk >= 70) return 'LOCK';
    final inside = price >= zLow && price <= zHigh;
    if (!inside) return 'WAIT';

    // TEST: zone 안쪽 진입
    // CONFIRM: 볼륨이 평균보다 증가 + 유리방향 신호/확률
    final vNow = s.candles.isNotEmpty ? s.candles.last.volume : 0.0;
    final vAvg = _avgVol(s.candles, 20);
    final volOk = vAvg > 0 ? (vNow / vAvg) >= 1.25 : false;
    final p = _pWin(s);
    final confOk = p >= 0.55;

    if (volOk && confOk) return 'CONFIRM';
    return 'TEST';
  }

  static double _pWin(FuState s) {
    final sp = (s.signalProb / 100.0);
    if (sp > 0) return sp.clamp(0.05, 0.95);
    return (s.confidence / 100.0).clamp(0.05, 0.95);
  }

  static double _avgVol(List<FuCandle> c, int n) {
    if (c.isEmpty) return 0.0;
    final start = math.max(0, c.length - n);
    final slice = c.sublist(start);
    final sum = slice.fold<double>(0.0, (a, b) => a + b.volume);
    return sum / slice.length;
  }

  static double _atr14(List<FuCandle> c) {
    if (c.length < 2) return 0.0;
    final n = math.min(14, c.length - 1);
    double sum = 0.0;
    for (int i = c.length - n; i < c.length; i++) {
      if (i <= 0) continue;
      final hi = c[i].high;
      final lo = c[i].low;
      final prevClose = c[i - 1].close;
      final tr = math.max(hi - lo, math.max((hi - prevClose).abs(), (lo - prevClose).abs()));
      sum += tr;
    }
    return sum / n;
  }

  static (double, double) _swingRange(List<FuCandle> c) {
    if (c.length < 5) return ( _minLow(c), _maxHigh(c) );
    // 간단 프랙탈: 최근 50개 중 local min/max
    final start = math.max(2, c.length - 50);
    double swingLow = double.infinity;
    double swingHigh = -double.infinity;
    for (int i = start; i < c.length - 2; i++) {
      final l = c[i].low;
      final h = c[i].high;
      if (l < c[i-1].low && l < c[i-2].low && l < c[i+1].low && l < c[i+2].low) {
        swingLow = math.min(swingLow, l);
      }
      if (h > c[i-1].high && h > c[i-2].high && h > c[i+1].high && h > c[i+2].high) {
        swingHigh = math.max(swingHigh, h);
      }
    }
    if (swingLow == double.infinity) swingLow = _minLow(c);
    if (swingHigh == -double.infinity) swingHigh = _maxHigh(c);
    return (swingLow, swingHigh);
  }

  static (double, double, double) _wickCluster(List<FuCandle> c) {
    if (c.isEmpty) return (0.0, 0.0, 0.0);
    final start = math.max(0, c.length - 60);
    final slice = c.sublist(start);

    // wick 길이 비율 평균: (wick / range)
    double wickiness = 0.0;
    final lows = <double>[];
    final highs = <double>[];
    for (final k in slice) {
      lows.add(k.low);
      highs.add(k.high);
      final range = (k.high - k.low).abs().clamp(1e-9, 1e18);
      final lowerWick = (math.min(k.open, k.close) - k.low).clamp(0.0, 1e18);
      final upperWick = (k.high - math.max(k.open, k.close)).clamp(0.0, 1e18);
      wickiness += ((lowerWick + upperWick) / range);
    }
    wickiness = (wickiness / slice.length).clamp(0.0, 2.0);

    lows.sort();
    highs.sort();
    final wickLow = lows[(lows.length * 0.10).floor().clamp(0, lows.length - 1)];
    final wickHigh = highs[(highs.length * 0.90).floor().clamp(0, highs.length - 1)];
    return (wickLow, wickHigh, wickiness / 2.0); // 0~1 근사
  }

  static double _minLow(List<FuCandle> c) {
    if (c.isEmpty) return 0.0;
    double m = c.first.low;
    for (final k in c) { if (k.low < m) m = k.low; }
    return m;
  }
  static double _maxHigh(List<FuCandle> c) {
    if (c.isEmpty) return 0.0;
    double m = c.first.high;
    for (final k in c) { if (k.high > m) m = k.high; }
    return m;
  }

  static double _zoneTouchScore(List<FuCandle> c, double zLow, double zHigh) {
    if (c.isEmpty) return 0.0;
    final start = math.max(0, c.length - 30);
    final slice = c.sublist(start);
    int touches = 0;
    for (final k in slice) {
      if (k.low <= zHigh && k.high >= zLow) touches++;
    }
    return (touches / slice.length).clamp(0.0, 1.0);
  }
}
